File: labs/02/src/Observable/Observable.php
<?php

namespace App\Observable;

use App\Observer\ObserverInterface;

abstract class Observable implements ObservableInterface
{
    protected array $observers = [];

    public function registerObserver(ObserverInterface $observer) : void
    {
        $this->observers[] = $observer;
    }

    public function notifyObservers() :void
    {
        $data = $this->getChangedData();

        foreach ($this->observers as $observer) {
            $observer->update($data);
        }
    }

    public function removeObserver(ObserverInterface $observer) : void
    {
        $key = array_search($observer, $this->observers, true);
        if($key){
            unset($this->observers[$key]);
        }
    }

    abstract protected function getChangedData() : \StdClass;
}

File: labs/09/lab9v3/src/common/observer/Observable.java
package common.observer;

import java.util.ArrayList;

public class Observable implements Observed {
    private ArrayList<Observer> observers;

    public Observable() {
        observers = new ArrayList<Observer>();
    }
    @Override
    public void registerObserver(Observer observer) {
        observers.add(observer);
    }

    public void notifyObservers() {
        observers.forEach(Observer::update);
    }
}
